### PR TITLE
#57838 Reschedule StatefulSet's Pods if nodeSelector has changed

### LINKED ISSUE #57838: Modifying nodeSelector on StatefulSet doesn't reschedule Pods
/kind bug

**What happened**:
Changing nodeSelector of a StatefulSet doesn't trigger rescheduling of it's existing pods. I `kubectl apply` the StatefulSet below, and wait for it's Pods to get scheduled onto Nodes with label `node_type: type1`. Then I change the nodeSelector label to `node_type: type2` and do `kubectl apply` again. 

```
apiVersion: apps/v1beta1
kind: StatefulSet
metadata:
  name: sstest
  labels:
    app: sstest
spec:
  replicas: 2
  serviceName: "service"
  template:
    metadata:
      labels:
        app: sstest
    spec:
      nodeSelector:
        node_type: type1
      containers:
            - name: nginx
              image: k8s.gcr.io/nginx-slim:0.8
```

**What you expected to happen**:

I expect the Pods to be rescheduled to the type2 Nodes, but nothing happens. The Pods only get rescheduled if I manually kill them using `kubectl delete pod sstest-0`.

**How to reproduce it (as minimally and precisely as possible)**:

1. Create a cluster in which some nodes have the label `node_type: type1` while some other nodes have the label `node_type: type2`.
2. Apply the above StatefulSet definition to the cluster.
3. Change the nodeSelector from `node_type: type1` to `node_type: type2` in the StatefulSet definition file.
4. Apply the file again.
5. Kill a Pod manually to verify that it gets rescheduled to another node.

**Anything else we need to know?**:

I tested the same scenario with Deployments. In that case the rescheduling worked as expected.

**Environment**:
- Kubernetes version (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"8", GitVersion:"v1.8.4", GitCommit:"9befc2b8928a9426501d3bf62f72849d5cbcd5a3", GitTreeState:"clean", BuildDate:"2017-11-20T05:28:34Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"8+", GitVersion:"v1.8.4-gke.1", GitCommit:"04502ae78d522a3d410de3710e1550cfb16dad4a", GitTreeState:"clean", BuildDate:"2017-12-08T17:24:53Z", GoVersion:"go1.8.3b4", Compiler:"gc", Platform:"linux/amd64"}
```

- Cloud provider or hardware configuration: GKE
- OS (e.g. from /etc/os-release): cos

### PR DESCRIPTION
**Pull request by antoniaklja**
**What this PR does / why we need it**:
Reschedule StatefulSet's Pods if nodeSelector has changed.

**Which issue(s) this PR fixes**:
Fixes #57838

**Special notes for your reviewer**:
none

**Release note**:
none

### CODE REVIEW

**Comment by antoniaklja** [comment]
@foxish @enisoc @dims @k82cn ready to review :)

**Comment by liggitt** [comment]
cc @kubernetes/sig-apps-pr-reviews @smarterclayton

File: pkg/controller/statefulset/stateful_pod_control.go [hunk 1]
```diff
@@ -106,6 +106,13 @@ func (spc *realStatefulPodControl) UpdateStatefulPod(set *apps.StatefulSet, pod
 				return err
 			}
 		}
+
+		// if the Pod does not conform to the StatefulSet's nodeSelector, update the Pod's nodeSelector
+		if !nodeSelectorMatches(set, pod) {
+			updateNodeSelector(set, pod)
```
Comment (k82cn) [inline pkg/controller/statefulset/stateful_pod_control.go]: In StatefulSet, it seems only update Pod's nodeSelector, who will trigger re-schedule?

Comment (antoniaklja) [reply pkg/controller/statefulset/stateful_pod_control.go]: please correct me if I'm wrong but in `pkg/controller/statefulset/stateful_pod_control.go:122` following line will commit the update:

`_, updateErr := spc.client.CoreV1().Pods(set.Namespace).Update(pod)`

does it answer for your question?

**Review by kow3ns — CHANGES_REQUESTED** [review]
I have concerns about this PR

1. This doesn't respect the StatefulSet ordering garauntees for termination and recreation. 
2. Even if it did, the implications of moving a StatefulSet Pod this way wrt to storage are not addressed. Particularly with respect to provisioning of zonal storage and the soon be released local PV feature.
3. Updating the node selector will not work as (I believe) you intend it to.
4. Most importantly, a change of this scope requires a design to be agreed upon by the community prior to implementation so that (1), (2), and (3) can be discussed.

**Comment by dhilipkumars** [comment]
Should this not trigger a rolling update instead? All the pods in this set are to be moved to a different set of nodes?

**Comment by antoniaklja** [comment]
> 1) This doesn't respect the StatefulSet ordering garauntees for termination and recreation.
I'm confused here.. because it follows the same way as `updateStorage` does.

> 2) Even if it did, the implications of moving a StatefulSet Pod this way wrt to storage are not addressed. Particularly with respect to provisioning of zonal storage and the soon be released local PV feature.
is there any additional logic in `updateStorage` needed?

> 3) Updating the node selector will not work as (I believe) you intend it to.
did I miss something in my code?

> 4) Most importantly, a change of this scope requires a design to be agreed upon by the community prior to implementation so that (1), (2), and (3) can be discussed.
@kow3ns can I help with that somehow?

@dhilipkumars looks like it will be easier and any complex logic is not necessary here, just triggering rolling update

File: pkg/controller/statefulset/stateful_pod_control.go (hunk 1 above)
Comment (k82cn) [reply pkg/controller/statefulset/stateful_pod_control.go]: nop; this code only update pod object in apiserver; no one will trigger 'rescheduler'.

**Comment by k82cn** [comment]
Agree with  kow3ns@ to draft a doc for more discussion; there's another requirements on avoiding re-scheduling  ( xref https://github.com/kubernetes/kubernetes/issues/57181 ).

File: pkg/controller/statefulset/stateful_pod_control.go [hunk 2]
```diff
@@ -106,6 +106,13 @@ func (spc *realStatefulPodControl) UpdateStatefulPod(set *apps.StatefulSet, pod
 				return err
 			}
 		}
+
+		// if the Pod does not conform to the StatefulSet's nodeSelector, update the Pod's nodeSelector
+		if !nodeSelectorMatches(set, pod) {
```
Comment (liggitt) [inline pkg/controller/statefulset/stateful_pod_control.go]: this breaks in the presence of admission that places additional constraints on the pods' nodeSelectors (e.g. https://kubernetes.io/docs/admin/admission-controllers/#podnodeselector)

**Pull request closed without merging**